### PR TITLE
Add a rule to check baseUri pattern

### PR DESCRIPTION
**Pull request by gkhunger-sf**
There should be a base uri and it should match this pattern:
https://{shortCode}.api.commercecloud.salesforce.com/test-family/test-api/{version}

### CODE REVIEW

File: test/mercury/base-uri-pattern.test.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+const PROFILE = "mercury";
+const RULE = "http://a.ml/vocabularies/data#base-uri-matches-pattern";
+
+describe("base uri pattern validation", () => {
+  let doc;
+
+  beforeEach(() => {
+    doc = getHappySpec();
+  });
+
```
Comment (unandyala) [inline test/mercury/base-uri-pattern.test.ts:25]: add test for happy path?

File: test/mercury/base-uri-pattern.test.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    breaksOnlyOneRule(result, RULE);
+  });
+
+  it("should not conform if the baseUri doesn't end with {version}", async () => {
+    doc["baseUri"] =
+      "https://{shortCode}.api.commercecloud.salesforce.com/test-family/test-api/{version}/";
+    const result = await validateFile(renderSpecAsFile(doc), PROFILE);
+
+    breaksOnlyOneRule(result, RULE);
+  });
```
Comment (kalagarraj) [inline test/mercury/base-uri-pattern.test.ts:135]: Nice number of tests for not conforming..

Can u pls add one for conforming the shortCode rule?

File: test/mercury/base-uri-pattern.test.ts (hunk 1 above)
Comment (gkhunger-sf) [reply test/mercury/base-uri-pattern.test.ts:26]: done

File: test/mercury/base-uri-pattern.test.ts (hunk 2 above)
Comment (gkhunger-sf) [reply test/mercury/base-uri-pattern.test.ts:156]: Added a happy path test

File: profiles/mercury.raml [hunk 3]
```diff
@@ -159,4 +160,14 @@ validations:
         pattern: ^v[0-9]+$
         minCount: 1
 
+  base-uri-matches-pattern:
+    message: The baseUri must match the pattern - https://{shortCode}.api.commercecloud.salesforce.com/<api-family>/<api-name>/{version}
+    targetClass: apiContract.WebAPI
+    functionConstraint:
+      code: |
+        function(resource) {
```
Comment (monkey-jeff) [inline profiles/mercury.raml:168]: Why code here instead of a pattern regex enforcement?

File: profiles/mercury.raml [hunk 4]
```diff
@@ -159,4 +160,14 @@ validations:
...
+          const baseUriRegex = /^https:\/\/{shortCode}.api.commercecloud.salesforce.com\/[a-z]+-?[a-z]+\/[a-z]+-?[a-z]+\/{version}$/;
+          const server = resource['apiContract:server'];
+          return server ? baseUriRegex.test(server[0]['core:urlTemplate'][0]) : false;
```
Comment (monkey-jeff) [inline profiles/mercury.raml:171]: server[0] seems a bit easy to break.  Can you have multiple servers specified?  Not sure we want to do this with custom code.

File: profiles/mercury.raml [hunk 5]
```diff
@@ -159,4 +160,14 @@ validations:
         pattern: ^v[0-9]+$
         minCount: 1
 
+  base-uri-matches-pattern:
+    message: The baseUri must match the pattern - https://{shortCode}.api.commercecloud.salesforce.com/<api-family>/<api-name>/{version}
+    targetClass: apiContract.WebAPI
+    functionConstraint:
+      code: |
+        function(resource) {
+          const baseUriRegex = /^https:\/\/{shortCode}.api.commercecloud.salesforce.com\/[a-z]+-?[a-z]+\/[a-z]+-?[a-z]+\/{version}$/;
```
Comment (monkey-jeff) [inline profiles/mercury.raml:169]: This does not allow a trailing / in the regex.  Is this intended?

File: test/mercury/base-uri-pattern.test.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      "https://{shortCode}.api.commercecloud.salesforce.com/test-family/test-api";
+
+    const result = await validateFile(renderSpecAsFile(doc), PROFILE);
+
+    breaksOnlyOneRule(result, RULE);
+  });
+
+  it("should not conform if the baseUri doesn't end with {version}", async () => {
+    doc["baseUri"] =
+      "https://{shortCode}.api.commercecloud.salesforce.com/test-family/test-api/{version}/";
```
Comment (monkey-jeff) [inline test/mercury/base-uri-pattern.test.ts:151]: I think this should probably be a success.

File: test/mercury/base-uri-pattern.test.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+  it("should not conform if the api family or the api name is missing", async () => {
+    doc["baseUri"] =
+      "https://{shortCode}.api.commercecloud.salesforce.com/test-api/{version}";
+
+    const result = await validateFile(renderSpecAsFile(doc), PROFILE);
+
+    breaksOnlyOneRule(result, RULE);
+  });
+
+  it("should not conform if the version is missing", async () => {
```
Comment (monkey-jeff) [inline test/mercury/base-uri-pattern.test.ts:140]: Should have tests with something like `{VERSION}` as well

File: test/mercury/base-uri-pattern.test.ts [hunk 8]
```diff
@@ -0,0 +1,157 @@
+/*
```
Comment (monkey-jeff) [inline test/mercury/base-uri-pattern.test.ts:1]: SHould try breaking your code some more here.  Do things like double `/`between segments.

File: profiles/mercury.raml (hunk 3 above)
Comment (gkhunger-sf) [reply profiles/mercury.raml:168]: To validate that a baseUri is part of the spec. `apiContract.urlTempalte` is nested inside `apiContract.Server`. When baseUri is not present in the raml, it doesn't even execute propertyConstraints.

File: profiles/mercury.raml (hunk 5 above)
Comment (gkhunger-sf) [reply profiles/mercury.raml:169]: Yes, it is intended. If we allow a trailing / we will end up with // when the baseUri is appended with endpoints.

Comment (monkey-jeff) [reply profiles/mercury.raml:169]: Followup...does anyone currently have a trailing / in their base uri.  If not this is fine.  If so I don't think we should force people to change it right now this close to GA.

Comment (gkhunger-sf) [reply profiles/mercury.raml:169]: Yeah, shopper-stores and shopper-search. I will remove the check.

**Pull request merged**